cpmpy/expressions/core.py

```diff
@@ -744,3 +744,196 @@ def _wsum_make(arg):
         return [-1], [arg.args[0]]
     # default
     return [1], [arg]
+
+
+# Custom dictionary and set classes for CPMpy expressions.
+# These do not inherit from python dict or set, because python abstracts from the underlying hash table.
+# Meaning we cannot access it ourselves.
+# This makes you rely on super.add, which is what we are trying to avoid..
+class cpm_dict:
+    def __init__(self):
+        # Maps hash values to lists of (key, value) pairs with that hash
+        self._hash_groups = {}
+
+    def same(self, key1, key2):
+        """
+        Custom key comparison logic (override this as needed).
+        Example: Compare keys as strings to allow "1" and 1 to coexist.
+        """
+        return str(key1) == str(key2)
```

Comment:
ideally first check on name being equal, which is faster? and then compare args directly?
return key1.name == key2.name and key1.args == key2.args

or you want to keep it generic for non Expressions too?
Still feels like calling the recursive string is excessive... any equality check can stop as soon as one component is different; while string needs to go all through before doing a comparison

Reply:
Comparing the strings still feels a bit ugly, and indeed we could fail faster, but in practice this should almost always return true (only will be false if we have a hash collision) so failing faster is probably not better if we slow down the check between identical expressions